### PR TITLE
Improve Json serialization by always reading a complete object.

### PR DESCRIPTION
**Pull request by brentschmaltz**
Basically, when reading a JsonToken read the entire token and position the reader at the next token.
When reading a JsonObject, the reader was positioned at a JsonTokenType.EndObject after reading.
If the runtime was currently reading a JsonObject, the exit condition of seeing a JsonTokenType.EndObject would be true and the serialization would exit. Same was true for JsonArray.

This will make it very difficult for embedded objects.

Tests were added to ensure that Arrays and Objects placed in the beginning, middle or end of the Json text will be processed correctly.

Test code was removed that was used to understand how the previous internal Newtonsoft, System.Text.Json and external Newtonsoft would work together. 
That code is no longer needed as we are dealing with issues as they arise.

### CODE REVIEW

File: test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs [hunk 1]
```diff
@@ -3,20 +3,287 @@
...
+        public static TheoryData<JsonSerializerTheoryData> RoundTripObjectsTheoryData
+        {
+            // there are a lot of tests here as this tests the lowest level of the serializer.
+            // Please do no use '.' in test names as double clicking on TheoryData.TestId will not copy the name to the clipboard.
+            // With so many tests conditional break points are very useful.
+            get
+            {
+                var theoryData = new TheoryData<JsonSerializerTheoryData>();
+                #region First LevelTests
+                // TODO - guid is serialized as a string make sure test exists for JsonWebToken to that TryGetValue<Guid> is tested
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs:67]: ?

File: test/Microsoft.IdentityModel.TestUtils/DataSets.cs [hunk 2]
```diff
@@ -213,20 +214,6 @@ public static JsonWebKey JsonWebKey2
                 ""x5t"":""kriMPdmBvx68skT8-mPAB3BseeA""
             }";
 
-        public static string JsonWebKeyMixedCaseString =>
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.TestUtils/DataSets.cs:216]: are we losing test coverage by removing this?

File: src/Microsoft.IdentityModel.Protocols.OpenIdConnect/Json/OpenIdConnectConfigurationSerializer.cs [hunk 3]
```diff
@@ -158,28 +156,27 @@ public static OpenIdConnectConfiguration Read(ref Utf8JsonReader reader, OpenIdC
...
                         reader.Read();
                         if (reader.TokenType == JsonTokenType.True)
-                            config.FrontchannelLogoutSupported = "True";
+                            config.FrontchannelLogoutSupported = JsonPrimitives.True;
                         else if (reader.TokenType == JsonTokenType.False)
-                            config.FrontchannelLogoutSupported = "False";
+                            config.FrontchannelLogoutSupported = JsonPrimitives.False;
                         else
-                            config.FrontchannelLogoutSupported = JsonPrimitives.ReadString(ref reader, MetadataName.FrontchannelLogoutSupported, ClassName, false);
+                            config.FrontchannelLogoutSupported = JsonPrimitives.ReadStringAsBool(ref reader, MetadataName.FrontchannelLogoutSessionSupported, ClassName, false);
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/Json/OpenIdConnectConfigurationSerializer.cs]: why are we reading from FrontchannelLogoutSessionSupported but writing to FrontchannelLogoutSupported?

File: test/Microsoft.IdentityModel.TestUtils/DataSets.cs (hunk 2 above)
Comment (brentschmaltz) [reply test/Microsoft.IdentityModel.TestUtils/DataSets.cs:216]: @keegan-caruso  i folded that with this change: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/2e4f4993b547c99d201577d65fbedd896981329a/test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/OpenIdConnectSerializationTests.cs#L24

File: test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs (hunk 1 above)
Comment (brentschmaltz) [reply test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs:67]: We have to make sure that is true.

File: src/Microsoft.IdentityModel.Protocols.OpenIdConnect/Json/OpenIdConnectConfigurationSerializer.cs (hunk 3 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Protocols.OpenIdConnect/Json/OpenIdConnectConfigurationSerializer.cs]: That needs to be changed, nice catch.
The error message would be off, but the deserialization correct.

File: src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs [hunk 4]
```diff
@@ -936,11 +1012,48 @@ public static void WriteObject(ref Utf8JsonWriter writer, string key, object obj
...
             else if (obj is float f)
+                // Below net6.0, we have to convert the float to a decimal otherwise values like 1.11 will be serailized as 1.11000001
+                // In failure cases, we will write the float as is.
+#if NET6_0_OR_GREATER
+                writer.WriteNumber(key, f);
+#else
+                #pragma warning disable CA1031 // Do not catch general exception types, we have seen TryParse fault.
+                try
+                {
+                if (decimal.TryParse(f.ToString(CultureInfo.InvariantCulture), out decimal dec))
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs]: nit: spacing

File: src/Microsoft.IdentityModel.Protocols.OpenIdConnect/Json/OpenIdConnectConfigurationSerializer.cs [hunk 5]
```diff
@@ -323,26 +321,25 @@ public static OpenIdConnectConfiguration Read(ref Utf8JsonReader reader, OpenIdC
                             else if (propertyName.Equals(MetadataName.EndSessionEndpoint, StringComparison.OrdinalIgnoreCase))
                                 config.EndSessionEndpoint = JsonPrimitives.ReadString(ref reader, propertyName, ClassName);
 
-                            // TODO these two properties are per spec 'boolean', we shipped 6x with them as string, if we change we may break folks.
-                            // probably best to mark the property obsolete with the gentle tag, then open up another property and keep them in sync,
-                            // remove the obsolete in 8.x
+                            // FrontchannelLogoutSessionSupported and FrontchannelLogoutSupported are per spec 'boolean'.
+                            // We shipped pervious versions accepting a string and transforming to a boolean.
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/Json/OpenIdConnectConfigurationSerializer.cs]:
```suggestion
                            // We shipped previous versions accepting a string and transforming to a boolean.
```

File: src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs [hunk 6]
```diff
@@ -24,6 +24,8 @@ internal static class JsonSerializerPrimitives
         // This number is the min between System.Text.Jsons default for
         // writing and reading max depth.
         const int MaxDepth = 64;
+        internal static string True ="true";
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs]: spacing?

File: test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs [hunk 7]
```diff
@@ -3,20 +3,287 @@
...
+                writer?.Dispose();
+            }
+
+            TestUtilities.AssertFailIfErrors(context);
+        }
+
+        public static TheoryData<JsonSerializerTheoryData> RoundTripObjectsTheoryData
+        {
+            // there are a lot of tests here as this tests the lowest level of the serializer.
+            // Please do no use '.' in test names as double clicking on TheoryData.TestId will not copy the name to the clipboard.
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs]:
```suggestion
            // Please do not use '.' in test names as double clicking on TheoryData.TestId will not copy the name to the clipboard.
```

File: test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs (hunk 1 above)
Comment (keegan-caruso) [reply test/Microsoft.IdentityModel.Tokens.Tests/Json/JsonSerializerPrimitivesTests.cs:67]: right, is it still a TODO?

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Comment by twogood** [comment]
I'm wishing for a release with this fix ASAP, been spending days trying to understand and workaround incorrect parsing of `"mtls_endpoint_aliases":{}` in `.well-known/openid-configuration`

**Pull request merged**